python/setup.py

```diff
@@ -82,18 +91,81 @@
+    return sys.platform == 'msys' or sys.platform == 'win32' and sys.version_string and 'GCC' in sys.version_string
+
+def makedirs(path, exist_ok):
+    # WARNING: path components to create should NOT include ".." components
+    # Documentation: https://docs.python.org/3/library/os.html#os.makedirs
+    if sys.version_info[:2] >= (3, 2):
+        os.makedirs(path, exist_ok=exist_ok)
+    else:
+        try:
+            os.makedirs(path)
+        except OSError as ex:
+            error_due_to_existing_dir = exist_ok and ex.errno == errno.EEXIST and os.path.isdir(path)
+            if not error_due_to_existing_dir:
+                raise
+
+def download(url):
+    return urllib_request.urlopen(url).read()
+
+def bazel_build(root_dir, build_ext=None):  # Keep this as a global function to make life easier during development
+    if is_native_windows_or_msys():
+        BUILD_SH = os.getenv('BUILD_SH')
+        SYSTEMROOT = os.getenv('SYSTEMROOT')
+        if (not BUILD_SH) and SYSTEMROOT and os.path.isfile(os.path.join(SYSTEMROOT, 'System32', 'bash.exe')):
+            msg = " ".join([
+                "You appear to have Bash from WSL, which Bazel is not compatible with.",
+                "To avoid potential problems, please explicitly set the %r environment variable for Bazel."
+            ]) % ('BUILD_SH',)
+            raise ValueError(msg)
+
+    assert os.path.isabs(root_dir), "need current working directory as an absolute path"
+
+    # TODO(mehrdadn): Is this directory needed anymore?
+    makedirs(os.path.join(root_dir, 'build'), exist_ok=True)
+
+    # Note: We are passing in sys.executable so that we use the same
+    # version of Python to build pyarrow inside the build.sh script. Note
+    # that certain flags will not be passed along such as --user or sudo.
+    # TODO(rkn): Fix this.
+    pyarrow_url = 'https://s3-us-west-2.amazonaws.com/arrow-wheels/3a11193d9530fe8ec7fdb98057f853b708f6f6ae/index.html'
+    subprocess.check_call([
+        sys.executable, '-m', 'pip', 'install', '-q', 'pyarrow==0.14.0.RAY',
+        '--find-links', pyarrow_url,
+        '--target', os.path.join(root_dir, 'python', 'ray', 'pyarrow_files')
+    ])
+    if (3, 6) <= sys.version_info[:2] <= (3, 7):
+        pickle5_backport_url = 'https://github.com/pitrou/pickle5-backport/archive/5186f9ca4ce55ae530027db196da51e08208a16b.zip'
+        pickle5_dir = os.path.join(root_dir, 'python', 'ray', 'pickle5_files')
+        work_dir = tempfile.mkdtemp()
+        try:
+            zipfile.ZipFile(io.BytesIO(download(pickle5_backport_url)), 'r').extractall(work_dir)
+            subprocess.check_call([sys.executable, 'setup.py', 'bdist_wheel'], cwd=work_dir)
+            for wheel in glob.glob(os.path.join(work_dir, 'dist', '*.whl')):
+                zipfile.ZipFile(wheel, 'r').extractall(pickle5_dir)
+        finally:
+            shutil.rmtree(work_dir)
+    bazel_env = dict(os.environ.copy(), **{'PYTHON%d_BIN_PATH' % sys.version_info[0]: sys.executable})
```

Comment:
a lot of python magic in this line. would be nicer if you just do `bazel_env = dict(os.environ.copy()` and then `bazel_env.update({"PYTHON_BIN_PATH": ...})`